src/Widgets/SharePopover.vala

```diff
@@ -20,146 +20,87 @@
-            hide ();
-        });
-
-        facebook_button.clicked.connect (() => {
-            try {
-                AppInfo.launch_default_for_uri ("https://www.facebook.com/sharer/sharer.php?u=%s".printf (uri), null);
-            } catch (Error e) {
-                warning ("%s", e.message);
-            }
-            hide ();
-        });
-
-        google_button.clicked.connect (() => {
-            try {
-                AppInfo.launch_default_for_uri ("https://plus.google.com/share?url=%s".printf (uri), null);
-            } catch (Error e) {
-                warning ("%s", e.message);
-            }
-            hide ();
-        });
-
-        twitter_button.clicked.connect (() => {
-            try {
-                AppInfo.launch_default_for_uri ("http://twitter.com/home/?status=%s %s".printf (body, uri), null);
-            } catch (Error e) {
-                warning ("%s", e.message);
-            }
-            hide ();
-        });
-
-        reddit_button.clicked.connect (() => {
-            try {
-                AppInfo.launch_default_for_uri ("http://www.reddit.com/submit?title=%s&url=%s".printf (body, uri), null);
-            } catch (Error e) {
-                warning ("%s", e.message);
-            }
-            hide ();
-        });
-
-        tumblr_button.clicked.connect (() => {
+    Gtk.Button make_share_button (string tooltip_text, string uri_template, string icon_name) {
+        var share_button = new Gtk.Button.from_icon_name (icon_name, Gtk.IconSize.DND);
+        share_button.tooltip_text = tooltip_text;
+        share_button.get_style_context ().add_class (Gtk.STYLE_CLASS_FLAT);
+        share_button.clicked.connect (() => {
             try {
-                AppInfo.launch_default_for_uri ("https://www.tumblr.com/share/link?url=%s".printf (uri), null);
+                // Make the text and the uri safe to include in an uri
+                string escaped_link = GLib.Uri.escape_string (shared_link);
+                string escaped_text = GLib.Uri.escape_string (shared_text);
+                string resulting_uri;
+                // Check if there's only one placeholder, if so, use it for a link
+                if (uri_template.index_of ("%s") == uri_template.last_index_of ("%s")) {
+                    resulting_uri = uri_template.printf (escaped_link);
+                } else {
+                    resulting_uri = uri_template.printf (escaped_text, escaped_link);
+                }
+                AppInfo.launch_default_for_uri (resulting_uri, null);
             } catch (Error e) {
                 warning ("%s", e.message);
             }
             hide ();
         });
-
-        telegram_button.clicked.connect (() => {
-            try {
-                AppInfo.launch_default_for_uri ("https://t.me/share/url?url=%s".printf (uri), null);
-            } catch (Error e) {
-                warning ("%s", e.message);
-            }
+        return share_button;
+    }
+        
```

Comment:
Instead of this helper function, make a private class "ShareButton" with this as a method and pass in the (sanitized) shared_text and shared_link properties on construction.